keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+We can think of evacuation as a managed and safer alternative to eviction.
+
+### Evacuation Instigator
+
+The evacuation instigator can be any entity in the system: node maintenance controller, descheduler,
+cluster autoscaler, or any application/controller interfacing with the affected application/pods
+(evacuee).
+
+The instigator's responsibility is to communicate an intent to a pod that it should be evacuated,
+according to the instigator's own internal rules. It should reconcile its intent in case the intent
+is removed by a third party. And it should remove its intent when the evacuation is no longer necessary.
+
+Example evacuation triggers:
+- Node maintenance controller: node maintenance triggered by an admin.
+- Descheduler: descheduling triggered by a descheduling rule.
+- Cluster autoscaler: node downscaling triggered by a low node utilization.
+
+It is understood that multiple evacuation instigators may request evacuation of the same pod at the
+same time. The instigators should coordinate their intent and not remove the evacuation until all
+instigators have dropped their intent.
+
+### Evacuee and Evacuator
+
+The evacuee can be any pod. There can be multiple evacuators for a single pod, and they should all
+advertise which pods they are responsible for. Normally, the owner/controller of the pod is the main
+evacuator. In a special case, the evacuee can be its own evacuator. The evacuator should decide what
+action to take when it observes an evacuation intent directed at that evacuator:
+1. It can reject the evacuation and wait for the pod to be evacuated by another evacuator or evicted
+   by the evacuation controller.
+2. It can do nothing and wait for the pod to be evacuated by another evacuator or evicted by the
+   evacuation controller. This is discouraged because it is slower than a simple rejection.
+3. It can start the evacuation and periodically respond to the evacuation intent to signal that the
+   evacuation is in progress and not stuck. Evacuation is at the discretion of the evacuator and
+   can take many forms:
+   - Migration of data (both persistent and ephemeral) from one node to another.
+   - Waiting for a cleanup and release of important resources held by the pod.
+   - Waiting for important computations to complete.
+   - Non-graceful deletion of the pod (`gracePeriodSeconds=0`).
+   - Deletion of a pod that is covered by a blocking PodDisruptionBudget. The controller of the
+     application should have additional logic to distinguish whether a disruption of a particular
+     pod will disrupt the application as a whole.
+   - After a partial cleanup (e.g. storage migrated, notification sent) and if the application is
+     still in an available state, the eviction API can be used to respect PodDisruptionBudgets.
+
+In the end, the evacuation should always end with a pod being deleted (evict or delete call) by one
```

Comment:
Early you said "The instigator's responsibility is to communicate an intent ... it should remove its intent when the evacuation is no longer necessary."  Reconcile that with "evacuation should always end with a pod being deleted" ?

Reply:
In this context (or in general), what is the difference of "pod termination" and "pod deletion"?

Reply:
thank you. so you mean by default, evacuation aims to delete, but there will be a separate set of rules for this mechanism to determine if a pod is going to be terminated instead?

Reply:
> thank you. so you mean by default, evacuation aims to delete,

yes

>  but there will be a separate set of rules for this mechanism to determine if a pod is going to be terminated instead?

I think it should be decided by the evacuator thought the Evacuation API if it is useful to keep the pod around after the termination.

Reply:
> Early you said "The instigator's responsibility is to communicate an intent ... it should remove its intent when the evacuation is no longer necessary." Reconcile that with "evacuation should always end with a pod being deleted" ?

I have updated this part and redefined the eviction completion as a pod termination (and not necessarily deletion); please see https://github.com/kubernetes/enhancements/pull/4565/commits/ed637b7af41fe9de842fe0d0909268b8d3cbb654